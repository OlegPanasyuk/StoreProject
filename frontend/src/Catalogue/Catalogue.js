import React, { Component } from 'react';
import { Container, Row, Col} from 'react-bootstrap';
import CatalogueMenu from '../CatalogueMenu/CatalogueMenu';
import BreadCrumbs from '../BreadCrumbs/BreadCrumbs';
import Goods from '../Goods/Goods';

/* The requesting to server needs push to single file(object) */

class Catalogue extends Component {
    constructor(props) {
        super(props);
        this.requestCatalogue = this.requestCatalogue.bind(this);
        this.rebiuldTree = this.rebiuldTree.bind(this);
        this.requestGoods = this.requestGoods.bind(this);
        this.obj = {};
        this.state = {
            items : {},
            itemsBreadCrumbs : [],
            itemsGoods : []
        };
    }

    fintNodeInTree(id, obj = this.obj) {
        
        for (let key in obj) {
            if (obj.hasOwnProperty(key)) {
                if (Number(key) === id) {
                    if(key===6) console.log(obj[key]);
                    return obj[key];
                } 
            }
        }
        for (let key in obj) { 
           
            if (obj[key]['children']) {
                this.fintNodeInTree(id, obj[key]['children']);
            }
        } 
    }

    sortItems(id, arr) {
        let self = this;
        let arrPushingElements  = arr.filter((el) => el.parent_id === id);
        arrPushingElements.forEach(element => {
            if (element.parent_id === -1) {
                self.obj[element.id_catalogue] = element;
                if (!self.obj[element.id_catalogue]['children']) {
                    self.obj[element.id_catalogue]['children'] = {};
                }
            } else {
                let parent = self.fintNodeInTree(element.parent_id);
                // if (element.parent_id === 6) console.log(parent, element, self.obj);
                if (parent) {
                    if (parent.children) {
                        parent.children[element.id_catalogue] = element;
                        if (!parent.children[element.id_catalogue]['children']) {
                            parent.children[element.id_catalogue]['children'] = {};
                        }
                    }
                }
            }
            self.sortItems(element.id_catalogue, arr);
        });
    }

    convertArrIntoObject(arr) {
        let self = this;
        arr.forEach((element) => {
            self.obj[element.id_catalogue] = element;
        })
    }

   
    requestCatalogue(id=null) {
        const XHR = new XMLHttpRequest();
        var self = this;
        let body = (id) ? `?id=${id}` : '';
        XHR.open('GET', 'http://localhost:3300/catalogue'+body, true);
        
        XHR.send();

        XHR.onreadystatechange = function() { // (3)
        if (XHR.readyState !== 4) return;
            if (XHR.status !== 200) {
                console.error( XHR.status, XHR.statusText);
            } else {
                let arr = [];
                arr.push( { 
                    id: null, 
                    name: 'Home'
                } );
                self.sortItems( (id) ? id : -1, JSON.parse( XHR.responseText )); 
                self.setState(() => ({
                    items : self.obj,
                    itemsBreadCrumbs : arr
                }));
            }
        }
    }

    requestGoods(catalogue_id = null, idgoods = null) {
        const XHR = new XMLHttpRequest();
        var self = this;
        let body = (catalogue_id) ? `?id_catalogue=${catalogue_id}` : '';
            body += (idgoods) ? `?id=${idgoods}` : '';
        XHR.open('GET', 'http://localhost:3300/goods'+body, true);
        
        XHR.send();

        XHR.onreadystatechange = function() { // (3)
        if (XHR.readyState !== 4) return;
            if (XHR.status !== 200) {
                console.error( XHR.status, XHR.statusText);
            } else {
                self.setState(() => ({
                    itemsGoods: JSON.parse( XHR.responseText )
                }));
                console.log('как то так',JSON.parse( XHR.responseText ));
            }
        }
    }
 

    componentDidMount() {
       this.requestCatalogue();
    }

    rebiuldTree(id = null) {
        if (id) {
            let parent = this.fintNodeInTree(Number(id));
            if (parent) {
                let { itemsBreadCrumbs } = this.state;
                let objToPush = {
                    id : parent.id_catalogue,
                    name : parent.name
                } ;
                let a = itemsBreadCrumbs.filter(el => el.id === id);
                
                if (a.length === 0) {
                    itemsBreadCrumbs.push( objToPush );
                }
                let children = (parent.children) ? parent.children : {};
                this.setState(() => ({
                    items : children,
                    itemsBreadCrumbs : itemsBreadCrumbs
                }));
            }
        } else {
            this.setState(() => ({
                items : this.obj,
                itemsBreadCrumbs : [{
                    id : null,
                    name : 'Home'
                }],
                itemsGoods : []
            }));
        }
        
    }

    
    render() {
        let { items, itemsBreadCrumbs, itemsGoods} = this.state;
        let obj = items;
        console.log('pre render',obj);
        return (
           <Container>
                <Row>
                    <Col sm='auto'>
                        <BreadCrumbs 
                            obj = { itemsBreadCrumbs } 
                            handleClick = { this.rebiuldTree }
                            >
                        </BreadCrumbs>
                        <CatalogueMenu 
                            obj = { items } 
                            handleClick = { this.rebiuldTree }
                            requestGoods = { this.requestGoods }
                            />
                    </Col>
                    <Col>
                        <Goods 
                            obj= { itemsGoods }
                            >

                        </Goods>
                    </Col>
                    
                </Row>
                        
            </Container>
        )
    }
}



export default Catalogue ;