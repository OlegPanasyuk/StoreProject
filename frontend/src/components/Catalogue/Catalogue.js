import React, { Component } from 'react';
import { Container, Row, Col } from 'react-bootstrap';
import PropTypes from 'prop-types';

/* The requesting to server needs push to single file(object) */

import CatalogueMenu from '../CatalogueMenu/CatalogueMenu';
import BreadCrumbs from '../BreadCrumbs/BreadCrumbs';
import Goods from '../Goods/Goods';
import CatalogueService from '../../Services/CatalogueService';
import GoodsService from '../../Services/GoodsService';


class Catalogue extends Component {
    constructor(props) {
        super(props);
        this.requestCatalogue = this.requestCatalogue.bind(this);
        this.rebuildTree = this.rebuildTree.bind(this);
        this.requestGoods = this.requestGoods.bind(this);
        this.obj = {};
        this.state = {
            items: {},
            itemsBreadCrumbs: [],
            itemsGoods: []
        };
    }


    componentDidMount() {
        this.requestCatalogue();
    }

    rebuildTree = (id = null) => {
        if (id) {
            const parent = this.findNodeInTree(Number(id));
            if (parent) {
                const { itemsBreadCrumbs } = this.state;
                const objToPush = {
                    id: parent.id_catalogue,
                    name: parent.name
                };
                const a = itemsBreadCrumbs.filter(el => el.id === id);

                if (a.length === 0) {
                    itemsBreadCrumbs.push(objToPush);
                }
                const children = (parent.children) ? parent.children : {};
                this.setState(() => ({
                    items: children,
                    itemsBreadCrumbs
                }));
            }
        } else {
            this.setState(() => ({
                items: this.obj,
                itemsBreadCrumbs: [{
                    id: null,
                    name: 'Home'
                }],
                itemsGoods: []
            }));
        }
    }

    convertArrIntoObject = (arr) => {
        arr.forEach((element) => {
            this.obj[element.id_catalogue] = element;
        });
    }

    requestCatalogue = (id = null) => {
        CatalogueService.getCatalogue(id).then((data) => {
            const arr = [];
            arr.push({
                id: null,
                name: 'Home'
            });
            this.sortItems(id || -1, data.data);
            this.setState(() => ({
                items: this.obj,
                itemsBreadCrumbs: arr
            }));
        });
    }

    // Needs to refactor logic of building tree
    findNodeInTree = (id, obj = this.obj) => {
        let element = null;
        Object.keys(obj).forEach((key) => {
            if (Object.prototype.hasOwnProperty.call(obj, key)) {
                if (Number(key) === id) {
                    element = obj[key];
                }
            }
        });
        Object.keys(obj).forEach((key) => {
            if (obj[key].children) {
                this.findNodeInTree(id, obj[key].children);
            }
        });
        return element;
    }

    sortItems = (id, arr) => {
        const self = this;
        const arrPushingElements = arr.filter(el => el.parent_id === id);
        arrPushingElements.forEach((element) => {
            if (element.parent_id === -1) {
                self.obj[element.id_catalogue] = element;
                if (!self.obj[element.id_catalogue].children) {
                    self.obj[element.id_catalogue].children = {};
                }
            } else {
                const parent = self.findNodeInTree(element.parent_id);
                if (parent) {
                    if (parent.children) {
                        parent.children[element.id_catalogue] = element;
                        if (!parent.children[element.id_catalogue].children) {
                            parent.children[element.id_catalogue].children = {};
                        }
                    }
                }
            }
            self.sortItems(element.id_catalogue, arr);
        });
    }

    requestGoods = (catalogueId = null, idgoods = null) => {
        GoodsService.getGoods({
            id: idgoods,
            id_catalogue: catalogueId
        }).then((data) => {
            if (data.status === 200) {
                this.setState(() => ({
                    itemsGoods: data.data
                }));
            }
        });
    }

    render() {
        const { items, itemsBreadCrumbs, itemsGoods } = this.state;
        const { addItemToBacket } = this.props;
        return (
            <Container>
                <Row>
                    <Col sm='auto'>
                        <BreadCrumbs
                            obj={itemsBreadCrumbs}
                            handleClick={this.rebuildTree}
                        />
                    </Col>
                </Row>
                <Row>
                    <Col sm='auto'>
                        <CatalogueMenu
                            obj={items}
                            handleClick={this.rebuildTree}
                            requestGoods={this.requestGoods}
                        />
                    </Col>
                    <Col>
                        <Goods
                            obj={itemsGoods}
                            addItemToBacket={addItemToBacket}
                        />
                    </Col>
                </Row>
            </Container>
        );
    }
}

Catalogue.propTypes = {
    addItemToBacket: PropTypes.func
};

Catalogue.defaultProps = {
    addItemToBacket: () => {}
};

export default Catalogue;
