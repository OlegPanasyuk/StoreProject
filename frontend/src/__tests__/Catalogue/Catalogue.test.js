import React from 'react';
import Catalogue from '../../Catalogue/Catalogue';
import { shallow, configure } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import Adapter from 'enzyme-adapter-react-16';


configure({ adapter: new Adapter() });

function setupComponent(props = {
   
}) {
    let component = shallow(<Catalogue {...props} />);
    return { component, props };
}

describe('Catalogue', () => {
    it('Component must be rendered', () => {
        let { component } = setupComponent();
        expect(shallowToJson(component)).toMatchSnapshot();
    });
});