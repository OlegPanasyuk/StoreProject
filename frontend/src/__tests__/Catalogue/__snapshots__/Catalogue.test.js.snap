// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Catalogue Component must be rendered 1`] = `
<Container
  as="div"
  fluid={false}
>
  <Bootstrap(Row)>
    <Col
      as="div"
      sm="auto"
    >
      <BreadCrumbs
        handleClick={[Function]}
        obj={Array []}
      />
    </Col>
  </Bootstrap(Row)>
  <Bootstrap(Row)>
    <Col
      as="div"
      sm="auto"
    >
      <CatalogueMenu
        handleClick={[Function]}
        obj={Object {}}
        requestGoods={[Function]}
      />
    </Col>
    <Col
      as="div"
    >
      <Goods
        obj={Array []}
      />
    </Col>
  </Bootstrap(Row)>
</Container>
`;
